File: desktop/src/com/inkostilation/pong/desktop/display/ObserverTree.java
package com.inkostilation.pong.desktop.display;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Tree;
import com.inkostilation.pong.desktop.notification.ClientNotifier;
import com.inkostilation.pong.notifications.IObserver;
import com.kotcrab.vis.ui.widget.VisTree;

public abstract class ObserverTree<A extends Actor, D> extends VisTree<ObserverTree.LobbyNode, A> implements IObserver<D> {

    public ObserverTree(Class<D> observableClass) {
        super();
        ClientNotifier.getInstance().subscribe(this, observableClass);
    }

    @Override
    public void observe(D... observable) {
        clearChildren();
        int pos = 0;
        for (D data: observable) {
            add(new LobbyNode(toActor(data), pos));
            pos++;
        }
    }

    abstract A toActor(D data);

    public static class LobbyNode extends Tree.Node {

        private int pos;

        public LobbyNode(Actor actor, int pos) {
            super(actor);
            this.pos = pos;
        }

        public int getPos() {
            return pos;
        }
    }
}

File: desktop/src/com/inkostilation/pong/desktop/display/ObserverTree.kt
package com.inkostilation.pong.desktop.display

import com.badlogic.gdx.scenes.scene2d.Actor
import com.inkostilation.pong.desktop.display.ObserverTree.LobbyNode
import com.inkostilation.pong.desktop.notification.ClientNotifier
import com.inkostilation.pong.notifications.IObserver
import com.kotcrab.vis.ui.widget.VisTree

abstract class ObserverTree<A : Actor?, D>(observableClass: Class<D>?) : VisTree<LobbyNode?, A>(), IObserver<D> {
    override fun observe(vararg observable: D) {
        clearChildren()
        var pos = 0
        for (data in observable) {
            add(LobbyNode(toActor(data), pos))
            pos++
        }
    }

    abstract fun toActor(data: D): A
    class LobbyNode(actor: Actor?, val pos: Int) : Node<Any?, Any?, Any?>(actor)

    init {
        ClientNotifier.Companion.getInstance().subscribe(this, observableClass)
    }
}